node/cn/gasprice/gasprice.go

```diff
@@ -155,3 +186,133 @@ func (gpo *Oracle) SuggestTipCap(ctx context.Context) (*big.Int, error) {
 		return unitPrice, nil
 	}
 }
+
+// suggestTipCapUsingFeeHistory returns a tip cap based on fee history.
+func (oracle *Oracle) suggestTipCapUsingFeeHistory(ctx context.Context) (*big.Int, error) {
+	head, _ := oracle.backend.HeaderByNumber(ctx, rpc.LatestBlockNumber)
+	headHash := head.Hash()
+
+	// If the latest gasprice is still available, return it.
+	oracle.cacheLock.RLock()
+	lastHead, lastPrice := oracle.lastHead, oracle.lastPrice
+	oracle.cacheLock.RUnlock()
+	if headHash == lastHead {
+		return new(big.Int).Set(lastPrice), nil
+	}
+	oracle.fetchLock.Lock()
+	defer oracle.fetchLock.Unlock()
+
+	// Try checking the cache again, maybe the last fetch fetched what we need
+	oracle.cacheLock.RLock()
+	lastHead, lastPrice = oracle.lastHead, oracle.lastPrice
+	oracle.cacheLock.RUnlock()
+	if headHash == lastHead {
+		return new(big.Int).Set(lastPrice), nil
+	}
+	var (
+		sent, exp int
+		number    = head.Number.Uint64()
+		result    = make(chan results, oracle.checkBlocks)
+		quit      = make(chan struct{})
+		results   []*big.Int
+	)
+	for sent < oracle.checkBlocks && number > 0 {
+		go oracle.getBlockValues(ctx, number, sampleNumber, result, quit)
+		sent++
+		exp++
+		number--
+	}
+	for exp > 0 {
+		res := <-result
+		if res.err != nil {
+			close(quit)
+			return new(big.Int).Set(lastPrice), res.err
+		}
+		exp--
+		// Nothing returned. There are two special cases here:
+		// - The block is empty
+		// - All the transactions included are sent by the miner itself.
+		// In these cases, use the latest calculated price for sampling.
+		if len(res.values) == 0 {
+			res.values = []*big.Int{lastPrice}
+		}
+		// Besides, in order to collect enough data for sampling, if nothing
+		// meaningful returned, try to query more blocks. But the maximum
+		// is 2*checkBlocks.
+		if len(res.values) == 1 && len(results)+1+exp < oracle.checkBlocks*2 && number > 0 {
+			go oracle.getBlockValues(ctx, number, sampleNumber, result, quit)
+			sent++
+			exp++
+			number--
+		}
+		results = append(results, res.values...)
+	}
+	price := lastPrice
+	if len(results) > 0 {
+		slices.SortFunc(results, func(a, b *big.Int) int { return a.Cmp(b) })
+		price = results[(len(results)-1)*oracle.percentile/100]
+	}
+	if price.Cmp(maxPrice) > 0 {
```

Comment:
1. Using `upperBoundBaseFee` for `maxPrice`
   The `maxPrice` is from geth, and it's for upper bound of the suggested priority fee (and suggested gasPrice before London HF). So it is not related to the upper bound of the dynamic basefee.

   However, we could use `upperBoundBaseFee` for the `maxPrice` - to limit the suggested priority fee never exceed the base fee?

2. Why use `maxPrice`?
   I've looked into geth code for reasons to introduce a upper bound for suggested priority fee, but could not find out. Seems like such concept was introduced a [long time ago](https://github.com/ethereum/go-ethereum/pull/990) when gas price oracle was first added, but couldn't find backgrounds for the upper limit.

   Instead of finding a suitable value or governance params for the `maxPrice`, why don't we just allow node operators to set `maxPrice` with cmd flags like [geth](https://github.com/ethereum/go-ethereum/blob/7bb3fb1481acbffd91afe19f802c29b1ae6ea60c/eth/gasprice/gasprice.go#L96) does? I couldn't find reasons to introduce `maxPrice` but also not having reasons not to use `maxPrice`. Since we are trying to enhance compatibility and improve UX for users familiar eth, we could bring such flag too.

Reply:
@blukat29 Sure, then I'll just leave `maxPrice` as constant value (500 ston) if @hyunsooda does not have any further comments.

In case we want to support node operators to change that value with cmd flag, we could make another PR for that change.

Reply:
Yes, they are not relevant, but I'm concerned about potential error cases where the suggested values return inaccurately due to the reliance on a hardcoded value, even when the correct answer should be higher

Having the same flag the Geth has sounded good to me, but it's okay, too, if you see that it's not mandatory.

Reply:
@hyunsooda @2dvorak I'm also good with adding the command line flags like in geth. gpo.blocks, gpo.percentile, gpo.maxprice. https://github.com/ethereum/go-ethereum/blob/v1.13.14/cmd/utils/flags.go#L791

Reply:
OK, I'll add command line flags for GPO configs.